src/plugins/validation/oas3/semantic-validators/security-definitions-ibm.js

```diff
@@ -0,0 +1,168 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+// Assertation 1: `type` is a necessary property and has four possible values: `apiKey`, `HTTP`, `oauth2`, `openIdConnect`
+// Assertation 2: name property is required for `apiKey` type
+// Assertation 3: `in` property is required for `apiKey` type, valid values are: `query`, `header` or `cookie`
+// Assertation 4: `scheme` property` is required for `http` type
+// Assertation 5: `flows` object is required for `oauth2` type
+// Assertation 6: `opedIdConnectUrl` property is required for `openIdConnect` and must be a valid url
+
+const validator = require('validator');
+
+module.exports.validate = function({ resolvedSpec }) {
+  const API_KEY = 'apiKey';
+  const OAUTH2 = 'oauth2';
+  const HTTP = 'http';
+  const OPENID_CONNECT = 'openIdConnect';
+  const authTypes = [API_KEY, HTTP, OAUTH2, OPENID_CONNECT];
+  const errors = [];
+  const warnings = [];
+  const securitySchemes =
+    resolvedSpec.components && resolvedSpec.components.securitySchemes;
+
+  for (const key in securitySchemes) {
+    const path = `securitySchemes.${key}`;
+    const security = securitySchemes[key];
+    const type = security.type;
+
+    if (!type) {
+      errors.push({
+        message: 'security scheme is missing required field `type`',
+        path
+      });
+    } else if (authTypes.indexOf(type) === -1) {
+      errors.push({
+        message:
+          '`type` must have one of the following types: `apiKey`, `oauth2`, `http`, `openIdConnect`',
+        path
+      });
+    } else {
+      //apiKey validation
+      if (type === API_KEY) {
+        const authIn = security.in;
+        if (!authIn || !['query', 'header', 'cookie'].includes(authIn)) {
+          errors.push({
+            message:
+              "apiKey authorization must have required 'in' property, valid values are 'query' or 'header' or 'cookie'.",
+            path
+          });
+        }
+        if (!security.name) {
+          errors.push({
+            message:
+              "apiKey authorization must have required 'name' string property. The name of the header or query property to be used.",
+            path
+          });
+        }
+      }
+      // oauth2 validation
+      else if (type === OAUTH2) {
+        const flows = security.flows;
+
+        if (!flows) {
+          errors.push({
+            message: "oauth2 authorization must have required 'flows' property",
+            path
+          });
+        } else if (
+          (flows.authorizationCode && !flows.authorizationCode.tokenUrl) ||
+          (flows.password && !flows.password.tokenUrl) ||
+          (flows.clientCredentials && !flows.clientCredentials.tokenUrl)
+        ) {
```

Comment:
I think structuring the code this way is not really what we want. I think its quite possible that an OAUTH security scheme could have a `flows` object that defines all three of `authorizationCode`, `password`, and `clientCredentials` flows. In that case, we want to validate that the `tokenUrl` is present in each one of them and issue a warning message for each case where it is not present. The current code will only issue a single message that `tokenURL` is missing without being specific to which one is in error.